From: Haoyang Chen <chenhaoyang@uniontech.com>
#Subject: [PATCH] winex11.drv: Minimizing a window with the owner property does not require a real minimization.
Message-Id: <21b9b2a1-cf7d-e7a9-d760-38b0857d3fe1@uniontech.com>
Date: Fri, 11 Dec 2020 17:33:22 +0800

From d3dcc0dec6c7e7262db521c360a3bb882021c3ef Mon Sep 17 00:00:00 2001
From: Haoyang Chen <chenhaoyang@uniontech.com>
Date: Fri, 11 Dec 2020 16:53:03 +0800
Subject: [PATCH] winex11.drv: Minimizing a window with the owner property does
 not require a real minimization.

Window minimization can get stuck if the window has the owner and
WS_POPUP attributes. This is because it has to be both on top of the
owner window and minimized.

Signed-off-by: Haoyang Chen <chenhaoyang@uniontech.com>
---
 dlls/winex11.drv/window.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -731,7 +731,8 @@ static void set_size_hints( struct x11drv_win_data *data, DWORD style )
     size_hints->flags |= PWinGravity;
 
     /* don't update size hints if window is not in normal state */
-    if (!(style & (WS_MINIMIZE | WS_MAXIMIZE)))
+    if ((X11DRV_get_whole_window( NtUserGetAncestor( NtUserGetWindowRelative( data->hwnd, GW_OWNER ), GA_ROOT ) ) && data->managed) ||
+            !(style & (WS_MINIMIZE | WS_MAXIMIZE)))
     {
         if (data->hwnd != NtUserGetDesktopWindow())  /* don't force position of desktop */
         {
@@ -2943,7 +2944,11 @@ void X11DRV_WindowPosChanged( HWND hwnd, HWND insert_after, UINT swp_flags,
         else if ((swp_flags & SWP_STATECHANGED) && (!data->iconic != !(new_style & WS_MINIMIZE)))
         {
             set_wm_hints( data );
-            data->iconic = (new_style & WS_MINIMIZE) != 0;
+            HWND owner_win = X11DRV_get_whole_window( NtUserGetAncestor( NtUserGetWindowRelative( data->hwnd, GW_OWNER ), GA_ROOT ));
+            if (owner_win && data->managed)
+                    sync_window_position( data, swp_flags, &old_window_rect, &old_whole_rect, &old_client_rect );
+
+            data->iconic = (new_style & WS_MINIMIZE) != 0 && !(owner_win && data->managed);
             TRACE( "changing win %p iconic state to %u\n", data->hwnd, data->iconic );
             if (data->iconic)
             {
@@ -3016,7 +3021,8 @@ UINT X11DRV_ShowWindow( HWND hwnd, INT cmd, RECT *rect, UINT swp )
     {
         if (((rect->left != -32000 || rect->top != -32000)) && hide_icon( data ))
         {
-            OffsetRect( rect, -32000 - rect->left, -32000 - rect->top );
+            if (!(X11DRV_get_whole_window( NtUserGetAncestor( NtUserGetWindowRelative( data->hwnd, GW_OWNER ), GA_ROOT ) ) && data->managed ))
+                OffsetRect( rect, -32000 - rect->left, -32000 - rect->top );
             swp &= ~(SWP_NOMOVE | SWP_NOCLIENTMOVE);
         }
         goto done;

-- 
2.40.0

